wgpu-hal/src/gles/egl.rs

```diff
@@ -608,8 +636,11 @@ impl Drop for Inner {
         {
             log::warn!("Error in destroy_context: {:?}", e);
         }
-        if let Err(e) = self.egl.instance.terminate(self.egl.display) {
-            log::warn!("Error in terminate: {:?}", e);
+
+        if decrement_display_reference_count(self.egl.display) {
+            if let Err(e) = self.egl.instance.terminate(self.egl.display) {
+                log::warn!("Error in terminate: {:?}", e);
+            }
         }
```

Comment:
This code is racy. After getting a decrement to hit zero, another thread might call `create` and bump the count up for that very display and call `egl.initialize` which then followed by this thread calling `terminate`, thus leaving the thread that just initialized with a broken display
(The logic doesn't work out like an an `arc` since in an `arc` there's a single defined creation via "new" whereas here we "new" on every reference count bump)

You'll have to do a lot more aggressive locking here: Decrement + `instance.terminate` and Increment + `egl.initialize` each have to be fully protected by the same mutex.

Reply:
:facepalm: nice catch, thank you!